File: cpp/questions/findRotation.cpp
#include <vector>
#include <array>

using namespace std;

class Solution {
private:
    static constexpr std::array<std::array<int, 2>, 4> dirs = {{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}};
public:
    bool findRotation(vector<vector<int>>& mat, vector<vector<int>>& target) {
        int n = mat.size();
        for (int t=0; t<4; ++t) {
            int tr = t<=1 ? 0 : n-1;
            int tc = 1<=t&&t<=2 ? n-1 : 0;
            int td = t;
            int md = 0;
            int mr = 0;
            int mc = 0;
            int l = 0;
            int p = 0;
            for (; p<n*n; ++p) {
                if (mat[mr][mc] != target[tr][tc]) {
                    break;
                }
                mr += dirs[md][0];
                mc += dirs[md][1];
                tr += dirs[td][0];
                tc += dirs[td][1];
                switch (md) {
                    case 0:
                        if (mc==n-l-1) {
                            md = 1;
                            td = (td + 1) % 4;
                        }
                        break;
                    case 1:
                        if (mr == n-l-1) {
                            md = 2;
                            td = (td + 1) % 4;
                        }
                        break;
                    case 2:
                        if (mc == l) {
                            md = 3;
                            td = (td + 1) % 4;
                        }
                        break;
                    case 3:
                        if (mr == 1+l) {
                            md = 0;
                            ++l;
                            td = (td + 1) % 4;
                        }
                        break;
                }
            }
            if (p == n*n) {
                return true;
            }
        }
        return false;
    }
};

File: python/questions/findRotation.py
from typing import List

class Solution:
    def findRotation(self, mat: List[List[int]], target: List[List[int]]) -> bool:
        n = len(mat)
        flag = (1 << 4) - 1
        for r in range(n):
            for c in range(n):
                flag &= (
                    (mat[r][c] == target[r][c]) |
                    ((mat[r][c] == target[c][n-1-r]) << 1) |
                    ((mat[r][c] == target[n-1-r][n-1-c]) << 2) |
                    ((mat[r][c] == target[n-1-c][r]) << 3)
                )
                if not flag:
                    return False
        return bool(flag)
